pkg/privilege/privileges/cache.go

```diff
@@ -1633,12 +1815,48 @@ func (p *MySQLPrivilege) getAllRoles(user, host string) []*auth.RoleIdentity {
 
 // Handle wraps MySQLPrivilege providing thread safe access.
 type Handle struct {
+	sctx sessionctx.Context
 	priv atomic.Pointer[MySQLPrivilege]
+	// Only load the active user's data to save memory
+	mu struct {
+		sync.RWMutex
+		activeUsers map[string]struct{}
+	}
 }
 
 // NewHandle returns a Handle.
-func NewHandle() *Handle {
-	return &Handle{}
+func NewHandle(sctx sessionctx.Context) *Handle {
+	var priv MySQLPrivilege
+	ret := &Handle{}
+	ret.sctx = sctx
+	ret.priv.Store(&priv)
+	ret.mu.activeUsers = make(map[string]struct{})
+	return ret
+}
+
+// ensureActiveUser ensure that the specific user data is loaded in-memory.
+func (h *Handle) ensureActiveUser(user string) error {
+	h.mu.RLock()
+	_, exist := h.mu.activeUsers[user]
+	h.mu.RUnlock()
+	if exist {
+		return nil
+	}
+
+	var data immutable
+	err := data.loadSomeUsers(h.sctx, user)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	old := h.Get()
+	h.priv.Store(old.merge(&data))
+
+	h.mu.Lock()
+	h.mu.activeUsers[user] = struct{}{}
```

Comment:
maybe the use case is more suitable for sync.Map, because the user key write once and read for multiple times, and user key accessing are disjoint.

Reply:
done